network/endpoint_linux.go

```diff
@@ -278,3 +284,120 @@ func deleteRoutes(interfaceName string, routes []RouteInfo) error {
 
 	return nil
 }
+
+// updateEndpointImpl updates an existing endpoint in the network.
+func (nw *network) updateEndpointImpl(existingEpInfo *EndpointInfo, targetEpInfo *EndpointInfo) (*endpoint, error) {
+	var ns *Namespace
+	var ep *endpoint
+	var err error
+
+	existingEpFromRepository := nw.Endpoints[existingEpInfo.Id]
+	log.Printf("[updateEndpointImpl] Going to retrieve endpoint with Id %+v to update.", existingEpInfo.Id)
+	if existingEpFromRepository == nil {
+		log.Printf("[updateEndpointImpl] Endpoint cannot be updated as it does not exist.")
+		err = errEndpointNotFound
+		return nil, err
+	}
+
+	netns := existingEpFromRepository.NetworkNameSpace
+	// Network namespace for the container interface has to be specified
+	if netns != "" {
+		// Open the network namespace.
+		log.Printf("[updateEndpointImpl] Opening netns %v.", netns)
+		ns, err = OpenNamespace(netns)
+		if err != nil {
+			return nil, err
+		}
+		defer ns.Close()
+
+		// Enter the container network namespace.
+		log.Printf("[updateEndpointImpl] Entering netns %v.", netns)
+		if err = ns.Enter(); err != nil {
+			return nil, err
+		}
+
+		// Return to host network namespace.
+		defer func() {
+			log.Printf("[updateEndpointImpl] Exiting netns %v.", netns)
+			if err := ns.Exit(); err != nil {
+				log.Printf("[updateEndpointImpl] Failed to exit netns, err:%v.", err)
+			}
+		}()
+	} else {
+		log.Printf("[updateEndpointImpl] Endpoint cannot be updated as the network namespace does not exist: Epid: %v", existingEpInfo.Id)
+		err = errNamespaceNotFound
+		return nil, err
+	}
+
+	log.Printf("[updateEndpointImpl] Going to update routes in netns %v.", netns)
+	if err = updateRoutes(existingEpInfo, targetEpInfo); err != nil {
+		return nil, err
+	}
+
+	// Create the endpoint object.
+	ep = &endpoint{
+		Id: existingEpInfo.Id,
+	}
+
+	// Update existing endpoint state with the new routes to persist
+	for _, route := range targetEpInfo.Routes {
+		ep.Routes = append(ep.Routes, route)
+	}
+
+	return ep, nil
+}
+
+func updateRoutes(existingEp *EndpointInfo, targetEp *EndpointInfo) error {
+	log.Printf("Updating routes for the endpoint %+v.", existingEp)
+	log.Printf("Target endpoint is %+v", targetEp)
+
+	existingRoutes := make(map[string]RouteInfo)
+	targetRoutes := make(map[string]RouteInfo)
+	var tobeDeletedRoutes []RouteInfo
+	var tobeAddedRoutes []RouteInfo
+
+	// we should not remove default route from container if it exists
+	// we do not support enable/disable snat for now
+	defaultDst := net.ParseIP("0.0.0.0")
+
+	for _, route := range existingEp.Routes {
```

Comment:
don't we get update route for default gateway?

Reply:
not in routes